### PR TITLE
Port openidconnect primitives tests

### LINKED ISSUE #817: Port the AspNet.Security.OpenIdConnect.Primitives tests
As part of 3.0, we'll need to port the ASOS primitives tests to OpenIddict: https://github.com/aspnet-contrib/AspNet.Security.OpenIdConnect.Server/tree/dev/test/AspNet.Security.OpenIdConnect.Primitives.Tests

The new primitives live here: https://github.com/openiddict/openiddict-core/tree/dev/src/OpenIddict.Abstractions/Primitives

Note: a few tests will need to be added to test the new `TryGetParameter` methods in `OpenIddictMessage` and `OpenIddictParameter`. Tests for `OpenIddictExtensions.GetResponseTypes()` will also be needed.

### PR DESCRIPTION
**Pull request by IgorHrabrov**
resolved #817

### CODE REVIEW

File: src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs [hunk 1]
```diff
@@ -23,10 +23,7 @@ public class OpenIddictConverter : JsonConverter
         /// <returns><c>true</c> if the type is supported, <c>false</c> otherwise.</returns>
         public override bool CanConvert([NotNull] Type type)
         {
-            if (type == null)
-            {
-                throw new ArgumentNullException(nameof(type));
-            }
+            type = type ?? throw new ArgumentNullException(nameof(type));
```
Comment (kevinchalet) [inline src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs]: I'm not a huge fan of this syntax, which also produces a different IL output (https://sharplab.io/#v2:C4LglgNgNAJiDUAfAAgJgIwFgBQyDMABGgQMI4DeOB1RhyALAQLIAUAysAE5gB2A5gQAOAQ07CAtgFNgkzgEoqNSthqqCYAGYEWIsVJmcCAXiMEeAVwgQFKtdWV27wABacA9gHczkrwEFOfOZSPMAAcpYQAKIAHgDGkoLAYG48LDwSkm4aOqIZBnJyANyKdgC+JQTlttQVFfhEjEyo7Fy8Arp5sjaqDo4d+rLGQrkDhgD8YwQu7l48PgT+gcFhETHxicmp6VJZOXrSXUUVVaVAA=)

Any chance you could revert these changes? This should help make your PR much smaller.

File: src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs [hunk 2]
```diff
@@ -120,7 +100,7 @@ public override void WriteJson([NotNull] JsonWriter writer, [NotNull] object val
                 {
                     writer.WritePropertyName(parameter.Key);
 
-                    var token = (JToken) parameter.Value;
+                    var token = (JToken)parameter.Value;
```
Comment (kevinchalet) [inline src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs]: Please revert styles changes. To make things easier for you, I can add an `.editorconfig` file. Would you be interested?

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs [hunk 3]
```diff
@@ -0,0 +1,872 @@
...
+        public void GetAcrValues_ReturnsExpectedAcrValues(string value, string[] values)
+        {
+            // Arrange
+            var request = new OpenIddictRequest
+            {
+                AcrValues = value
+            };
+
+            // Act
+            var actualValues = request.GetAcrValues();
```
Comment (kevinchalet) [inline test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:46]: Are these changes required because we now return immutable collections, where the order is not respected?

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs [hunk 4]
```diff
@@ -0,0 +1,872 @@
...
+            // Arrange
+            var request = new OpenIddictRequest
+            {
+                ResponseType = type
+            };
+
+            // Act and assert
+            Assert.Equal(result, request.IsAuthorizationCodeFlow());
+        }
+
```
Comment (kevinchalet) [inline test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:473]: Not necessarily for this PR, but we'll need new tests for the extensions introduced as part of the device flow support (mostly a note for myself, but let me know if you're interested in doing it)

**Review by kevinchalet — CHANGES_REQUESTED** [review]

**Comment by kevinchalet** [comment]
Woot, thanks for your contribution @IgorHrabrov! 👏

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs (hunk 3 above)
Comment (kevinchalet) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:46]: I opened https://github.com/openiddict/openiddict-core/issues/846 to track future improvements to revert to ASOS' behavior, where the returned collections respected the order set by the client app.

File: src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs (hunk 1 above)
Comment (IgorHrabrov) [reply src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs]: Rollbacking changes

File: src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs (hunk 2 above)
Comment (IgorHrabrov) [reply src/OpenIddict.Abstractions/Primitives/OpenIddictConverter.cs]: Rollbacking changes

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs (hunk 3 above)
Comment (IgorHrabrov) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:46]: Yes, ImmutableHashSet<T> is unordered set, what  failed unit tests and that's why I changed the tests

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs (hunk 4 above)
Comment (IgorHrabrov) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:473]: If it's small changes I can add them here, just let me know what methods to cover, otherwise it may be better to do in separate PR

Comment (kevinchalet) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:473]: Only 1 extension and 4 properties were needed for the device flow, so it should be trivial:

https://github.com/openiddict/openiddict-core/blob/af1fb2e8907b3dce6d50558a6b90874a6680837b/src/OpenIddict.Abstractions/Primitives/OpenIddictExtensions.cs#L474-L488

https://github.com/openiddict/openiddict-core/blob/af1fb2e8907b3dce6d50558a6b90874a6680837b/src/OpenIddict.Abstractions/Primitives/OpenIddictRequest.cs#L195-L202

https://github.com/openiddict/openiddict-core/blob/af1fb2e8907b3dce6d50558a6b90874a6680837b/src/OpenIddict.Abstractions/Primitives/OpenIddictRequest.cs#L420-L427

https://github.com/openiddict/openiddict-core/blob/af1fb2e8907b3dce6d50558a6b90874a6680837b/src/OpenIddict.Abstractions/Primitives/OpenIddictResponse.cs#L87-L94

https://github.com/openiddict/openiddict-core/blob/af1fb2e8907b3dce6d50558a6b90874a6680837b/src/OpenIddict.Abstractions/Primitives/OpenIddictResponse.cs#L177-L184

Comment (IgorHrabrov) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:473]: Ok, I will cover this

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs (hunk 3 above)
Comment (kevinchalet) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:46]: Great, I'll update the tests once we move back to collection types that respect the user-defined order.

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictConverterTests.cs [hunk 5]
```diff
@@ -0,0 +1,322 @@
...
+        [InlineData(typeof(OpenIddictParameter?))]
+        [InlineData(typeof(OpenIddictParameter[]))]
+        [InlineData(typeof(OpenIddictParameter?[]))]
+        [InlineData(typeof(object))]
+        [InlineData(typeof(long))]
+        public void ReadJson_ThrowsAnExceptionForUnsupportedType(Type type)
+        {
+            // Arrange
+            var converter = new OpenIddictConverter();
+            using var reader = new JsonTextReader(new StringReader(@"{""name"":""value""}"));
```
Comment (kevinchalet) [inline test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictConverterTests.cs:113]: I like that. Should we add `using` in the other tests instantiating `JsonTextReader`? (at least for consistency)

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs (hunk 4 above)
Comment (IgorHrabrov) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictExtensionsTests.cs:473]: Done

File: test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictConverterTests.cs (hunk 5 above)
Comment (IgorHrabrov) [reply test/OpenIddict.Abstractions.Tests/Primitives/OpenIddictConverterTests.cs:113]: I don't think it's necessary right now. I faced only 2 places with warnings (which I fixed)

**Comment by kevinchalet** [comment]
Merged, thanks for your contribution! 👏
